### PR TITLE
feat: Make `inline_function` work on methods

### PR DESCRIPTION
**Pull request by Veykril**
Now called `inline_call`.

### CODE REVIEW

File: crates/ide_assists/src/handlers/inline_function.rs [hunk 1]
```diff
@@ -2,14 +2,64 @@ use ast::make;
...
-    AstNode,
+    ted, AstNode,
 };
 
 use crate::{
     assist_context::{AssistContext, Assists},
     AssistId, AssistKind,
 };
 
+// Assist: inline_method
```
Comment (matklad) [inline crates/ide_assists/src/handlers/inline_function.rs:13]: From the user’s perspective, I think there should be only one assist: “inline call”, which works for function and method calls.

And some day we’d want to create “inline function” that is calllable at the der-site, and inlines all usages.

File: crates/syntax/src/ast/make.rs [hunk 2]
```diff
@@ -330,19 +337,18 @@ pub fn arg_list(args: impl IntoIterator<Item = ast::Expr>) -> ast::ArgList {
...
-}
-pub fn ident_mut_pat(name: ast::Name) -> ast::IdentPat {
-    return from_text(&name.text());
-
-    fn from_text(text: &str) -> ast::IdentPat {
-        ast_from_text(&format!("fn f(mut {}: ())", text))
+    if mut_ {
+        s.push_str("mut ");
     }
+    let _ = write!(s, "{}", name);
```
Comment (matklad) [inline crates/syntax/src/ast/make.rs:349]: We have format_to! for this

File: crates/ide_assists/src/handlers/extract_function.rs [hunk 3]
```diff
@@ -341,9 +341,9 @@ impl Param {
         let var = self.var.name(ctx.db()).unwrap().to_string();
         let var_name = make::name(&var);
         let pat = match self.kind() {
-            ParamKind::MutValue => make::ident_mut_pat(var_name),
+            ParamKind::MutValue => make::ident_pat(false, true, var_name),
             ParamKind::Value | ParamKind::SharedRef | ParamKind::MutRef => {
-                make::ident_pat(var_name)
+                make::ext::simple_ident_pat(var_name)
```
Comment (matklad) [inline crates/ide_assists/src/handlers/extract_function.rs:346]: 👍

File: crates/syntax/src/ast/make.rs [hunk 4]
```diff
@@ -330,19 +337,18 @@ pub fn arg_list(args: impl IntoIterator<Item = ast::Expr>) -> ast::ArgList {
     ast_from_text(&format!("fn main() {{ ()({}) }}", args.into_iter().format(", ")))
 }
 
-pub fn ident_pat(name: ast::Name) -> ast::IdentPat {
-    return from_text(&name.text());
-
-    fn from_text(text: &str) -> ast::IdentPat {
-        ast_from_text(&format!("fn f({}: ())", text))
+pub fn ident_pat(ref_: bool, mut_: bool, name: ast::Name) -> ast::IdentPat {
```
Comment (matklad) [inline crates/syntax/src/ast/make.rs:340]: Two bool args look awkward at the call site, but I don’t see an obviously better soln

**Comment by Veykril** [comment]
There is a lot of improvement potential for the assist here still.

 - For one better handling of the self param, which currently always emits `ref (mut)` for `&(mut )self` params even if not necessary.
 - A problem we currently have is that function parameters can have coercions happen, which this assist will lose as it doesnt emit type ascriptions for the let statements.
 - And there is the general potential for just not even generating some of the let statements and inline the expressions instead depending on the expression(I'll work on this next).

bors r+

**Comment by flodiebold** [comment]
We should start recording adjustments (for coercions and self parameter autoref / deref) in type inference sooner rather than later so we can handle these cases more correctly!

**Comment by Veykril** [comment]
Definitely, I think this is the third issue/feature which wants coercions tracked. I'll open a new issue for this one to track them together.

**Pull request merged**